<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Users Guide for R/funqtl}
-->

# funqtl: QTL mapping with function-valued traits

```{r knitr_options, echo=FALSE, results=FALSE}
library(knitr)
opts_chunk$set(fig.width = 12)
```

```{r loading_run, echo=FALSE, message=FALSE}
library(funqtl)
data(simspal)
```

This vignette illustrates the use of the funqtl package for QTL
mapping with function-valued traits (e.g., growth measured over time).


## Data

We will consider the analysis of a simulated data set, `simspal`.
There are `r nind(simspal)` recombinant inbred lines typed at a total
of `r totmar(simspal)` markers on `r nchr(simspal)` chromosomes.

We first load the funqtl package (which will also load R/qtl and other
packages) and the data.

```{r loading, eval=FALSE}
library(funqtl)
data(simspal)
```


Here's a quick summary of the data.

```{r summary_simspal}
summary(simspal)
```


The following plots the genetic marker map and the function-valued trait
for five of the RIL.

```{r plot_map_and_pheno}
par(mfrow=c(1,2))
plotMap(simspal, main="")
plot(1:241, simspal$pheno[160,], type="l", xlab="Time",
	  ylim=c(-120,0), ylab="Root Tip Angle (degrees)")
ind <- c(19, 20, 132, 72)
color <- c("blue", "red", "green", "orange")
for(i in seq(along=ind))
  lines(1:241, simspal$pheno[ind[i],], col=color[i])
```



## Single-QTL analysis at the individual time points

We first perform single-QTL genome scans at each time point, individually.

We use `calc.genoprob` in R/qtl to calculate QTL genotype
probabilities and then `scanone` to perform the genome scans, using
Haley-Knott regression (Haley and Knott 1992). We'll perform
calculations solely at the marker positions (`step=0`) to speed up the
calculations. We'll also consider only every fifth time point.

```{r scanone}
phe <- seq(1, nphe(simspal), by=5)
simspal <- calc.genoprob(simspal, step=0)
out <- scanone(simspal, pheno.col = phe, method="hk")
```

The function `geteffects` estimates the QTL effect at each locus, for
each time point.

```{r geteffects}
eff <- geteffects(simspal, pheno.cols=phe)
```

The function `plotlod` plots a heat map of signed LOD scores: the LOD
scores, taking the signs of the estimated effects.

```{r plotload, fig.height=10, fig.width = 12}
plotlod(out, eff, phe, gap=15,
	    main="The LOD image of the simspal data set",
        ylab="Time")
```

The x-axis represents genomic position and the y-axis represents
time, and so each horizontal slice is a genome scan for one time
point. We plot a signed LOD score, with the sign representing the
estimated direction of the QTL effect. The most prominant QTL are on
chromosomes 1 and 4.

The chromosome 1 QTL affects later times, and the chromosome 4
allele affects earlier times. There is an additional QTL of interest on distal chromosome 3

## SLOD and MLOD scores

```{r slodmlod}
out1 <- scanoneF(simspal, pheno.cols = 1:241, method="hk")

```
The SLOD and MLOD statistics combine the results across time points, by taking the average
or the maximum LOD, respectively, at each genomic location.


```{r slodmlodcurve, fig.height=10, fig.width = 12}
par(mfrow=c(2,1))
plot(out1, ylim=c(0,3.5), main="The SLOD curve for simspal data",
	 bandcol="gray90")
abline(h=2.02, col="red", lty=3)
plot(out1, lodcolumn=2, ylim=c(0,7),
     main="The MLOD curve for simspal data", bandcol="gray90")
abline(h=3.46, col="red", lty=3)

# permutation threshold
# permout <- scanoneF(simspal, pheno.cols=1:241,
#                     method = "hk", n.perm=1000)
# display 5, 10 % threshold of permutation result
# summary(permout)
```

The results are in Figure above. Horizontal lines indicate the 5% genome-wide significance thresholds, derived by a permutation test. We didn't run the code above since it takes a long time (about one hour maybe) .

## Getting multiple QTL

```{r stepwiseqtlscan}

#qtlslod <- stepwiseqtlF(simspal, pheno.cols = 1:241,
#                   max.qtl = 6, usec = "slod",
#                   method = "hk",
#                   penalties = c(2.02, 2.62, 1.74) )
simspal <- calc.genoprob(simspal, step=0)
qtlslod <- makeqtl(simspal, chr = c(1, 4),
               pos = c(36.6, 27.8), what = "prob")
```

`stepwiseqtlF` function returns qtl object by using stepwise QTL selection. It takes long time to run this function, so I run `makeqtl` function to get the result of `stepwiseqtlF` function. You will get the same result if you run `stepwiseqtlF` function.


``` {r profilelodimage, fig.height = 10, fig.width = 14}
lodmat1.c <- getprofile(simspal, qtl =  qtlslod, pheno.cols = phe,
                             formula = y~Q1 + Q2 , method = "hk",
							 verbose = F, tpy="comb")
plotprofile(lodmat1.c, mval = 8, col=heat.colors(100)[100:1],
	        main="The Profile LOD image of data")
```
You can get profilelod image by using `getprofile` function and `plotprofile` function. You need to have an option `step = 0` in `calc.genoprob` function.

``` {r profile, fig.height = 10, fig.width = 14}
refqtlslod <- refineqtlF(simspal, pheno.cols = 1:241,
                         usec = "slod", qtl= qtlslod,
						 method = "hk", keeplodprofile = T)

plotLodProfile(refqtlslod)
```

You can also make slod profile(or mlod profile) curve by using `refineqtlF` and `plotLodProfile` function.

## Effect plot

```{r effectplot}

slodeff <- vector("list", nphe(simspal))

for(i in 1:nphe(simspal)) {
    slodeff[[i]] <- summary(fitqtl(simspal, phe=i, qtl=qtlslod,
                            method="hk", get.ests=TRUE,
							dropone=FALSE))$ests[,1]*c(1,2,2)
}

nam <- names(slodeff[[1]])
slodeff <- matrix(unlist(slodeff), byrow=TRUE, ncol=length(nam))
colnames(slodeff) <- nam

time <- (0:240)/30
```
To further characterize the effects of the QTL in the context of the inferred multiple-QTL
models, we fit the selected multiple-QTL models at each time point, individually. the estimated baseline function and the estimated QTL effects, as a function of time.
The estimated QTL effects in panels are for the difference between two alleles.


```{r effplot2, fig.height=5, fig.width = 14}
par(mfrow=c(1,3))
plot(time, slodeff[,1], lwd=2, type="l",
	xlab="Time (hours)",
	ylab="Tip angle (degrees)", col="red", ylim=c(-110,0))
	mtext("baseline curve", side=3, line=0.5)

plot(time, slodeff[,2], lwd=2, ylim = c(-5,9), type="l",
     xlab="Time (hours)",
	 ylab="QTL effect", col="red")
	abline(h=0)
	mtext("chr 1, 37 cM", side=3, line=0.5)

plot(time, slodeff[,3], lwd=2, ylim = c(-5,9), type="l",
     xlab="Time (hours)",
	ylab="QTL effect", col="red")
	mtext("chr 4, 28 cM", side=3, line=0.5)
	abline(h=0)
```


## HKLOD, FLOD, SL and ML scores

```{r cv}
cvout <- cvfold(simspal, basisset = 15:60, fold = 10, random = F)

```

We need to decide how many basis functions to be used by doing cross
validation. We can do this using `cvfold` function. `basisset` option
describe the number of basis sets to be tested. `fold` indicate how
many folders to be used in cross validation. In this code, we divide
our time points with equily spaced 10 sets. We used 9 folder to evaluate the
functional. And calculated sum of squared errors for the left one
folder.



```{r cvfig, fig.height=10, fig.width = 12}
plot(15:60, cvout, xlab = "The number of basis",
     main = "sum of squared errors", type = "l")

```
We decided to use 41 basis functions from the Figure above.


```{r hklodttc2}
Y <- calcfunpca(simspal, criteria=0.99, n.max = 8, nbasis = 41)

outhk <- scanoneM(simspal, Y = Y, method="hk")
outf <- scanoneM(simspal, Y = Y, method="f")
outsl <- scanoneM(simspal, Y = Y, method="sl")
outml <- scanoneM(simspal, Y = Y, method="ml")

```
The SLOD and MLOD statistics combine the results across time points, by taking the average
or the maximum LOD, respectively, at each genomic location.


```{r slodmlodcurve2, fig.height=10, fig.width = 12}
par(mfrow=c(4,1))
plot(outhk, ylim=c(0,8), main="The HKLOD curve for simspal data",
     bandcol="gray90")
abline(h=4.32, col="red", lty=3)
plot(outf, ylim=c(0,3.5), main="The FLOD curve for simspal data",
     bandcol="gray90")
abline(h=2.16, col="red", lty=3)
plot(outsl, ylim=c(0,2), main="The SL curve for simspal data",
     bandcol="gray90")
abline(h=1.05, col="red", lty=3)
plot(outml, ylim=c(0,4), main="The ML curve for simspal data",
     bandcol="gray90")
abline(h=3.05, col="red", lty=3)

# permutation threshold
# permouthk <- scanoneF(simspal, Y=Y, method = "hk", n.perm=1000)
# permoutf <- scanoneF(simspal, Y=Y, method = "f", n.perm=1000)
# permoutsl <- scanoneF(simspal, Y=Y, method = "sl", n.perm=1000)
# permoutml <- scanoneF(simspal, Y=Y, method = "ml", n.perm=1000)
# summary(permouthk) # display 5, 10 % threshold
# summary(permoutf) # display 5, 10 % threshold
# summary(permoutsl) # display 5, 10 % threshold
# summary(permoutml) # display 5, 10 % threshold
```

The results are in Figure above. Horizontal lines indicate the 5%
genome-wide significance thresholds, derived by a permutation test. We
didn't run the code (for permutation test) above since it takes a long time (about one hour maybe) .

## Getting multiple QTL

```{r stepwiseqtlscan2}


#qtlhklod <- stepwiseqtlM(simspal, Y = Y, max.qtl = 6,
#                   method = "hk", penalties = c(4.32, 6, 6) )

#qtlflod <- stepwiseqtlM(simspal, Y = Y, max.qtl = 6,
#                   method = "f", penalties = c(2.16, 6, 6) )

#qtlsllod <- stepwiseqtlM(simspal, Y = Y, max.qtl = 6,
#                   method = "sl", penalties = c(1.05, 6, 6) )

#qtlmllod <- stepwiseqtlM(simspal, Y = Y, max.qtl = 6,
#                   method = "ml", penalties = c(3.05, 6, 6) )

simspal <- calc.genoprob(simspal, step=0)
qtlhklod <- makeqtl(simspal, chr = c(1, 4),
               pos = c(62.3, 40.3), what = "prob")
qtlsllod <- makeqtl(simspal, chr = c(1, 4),
               pos = c(62.3, 40.3), what = "prob")
```

`stepwiseqtlM` function returns qtl object by using stepwise QTL
selection. It takes about 2~3 minute, so I run `makeqtl` function to get the result of `stepwiseqtlM` function. You will get the same result if you run `stepwiseqtlM` function.

``` {r profilecode, fig.height = 8, fig.width = 14}
 refqtlhklod <- refineqtlM(simspal, Y=Y, method = "hk",
                           qtl= qtlhklod, keeplodprofile = T)
 refqtlsllod <- refineqtlM(simspal, Y=Y, method = "sl",
                           qtl= qtlhklod, keeplodprofile = T)

par(mfrow=c(1,2))
plotLodProfile(refqtlhklod)
plotLodProfile(refqtlsllod)
```

You can also make hklod profile(or flod, sl and ml profile) curve by using
`refineqtlM` and `plotLodProfile` function. Here's HKLOD profile curve
and SL profile curves. Both penalized-HKLOD and penalized-SL found
2-QTL model. One at 62.3cM of chromosome 1 and another at 40.3cM of
chromosome 4.

## Effect plot

```{r effectplot2}

slodeff <- vector("list", nphe(simspal))

for(i in 1:nphe(simspal)) {
    slodeff[[i]] <- summary(fitqtl(simspal, phe=i, qtl=qtlhklod,
                            method="hk", get.ests=TRUE,
							dropone=FALSE))$ests[,1]*c(1,2,2)
}

nam <- names(slodeff[[1]])
slodeff <- matrix(unlist(slodeff), byrow=TRUE, ncol=length(nam))
colnames(slodeff) <- nam

time <- (0:240)/30
```
To further characterize the effects of the QTL in the context of the inferred multiple-QTL
models, we fit the selected multiple-QTL models at each time point, individually. the estimated baseline function and the estimated QTL effects, as a function of time.
The estimated QTL effects in panels are for the difference between two alleles.


```{r effplot3, fig.height=5, fig.width = 14}
par(mfrow=c(1,3))
plot(time, slodeff[,1], lwd=2, type="l",
	xlab="Time (hours)",
	ylab="Tip angle (degrees)", col="red", ylim=c(-110,0))
	mtext("baseline curve", side=3, line=0.5)

plot(time, slodeff[,2], lwd=2, ylim = c(-5,9), type="l",
     xlab="Time (hours)",
	 ylab="QTL effect", col="red")
	abline(h=0)
	mtext("chr 1, 62 cM", side=3, line=0.5)

plot(time, slodeff[,3], lwd=2, ylim = c(-5,9), type="l",
     xlab="Time (hours)",
	ylab="QTL effect", col="red")
	mtext("chr 4, 40.3 cM", side=3, line=0.5)
	abline(h=0)
```

