<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Users Guide for R/funqtl}
-->

# QTL mapping on Simulated data sets

This is a simple analysis on simulated data set, simspal, using R/funqtl package.
  There are 162 RIL individuals typed at a total of 234 markers on five
  chromosomes.


```{r loading}
library(qtl)
library(funqtl)
data(simspal)
simspal
```


## Data

```{r plots, fig.width = 12}
pheno1 <- simspal$pheno
map10 <- pull.map(simspal)
par(mfrow=c(1,2))
plot(map10, main="")
plot(1:241, pheno1[160,], type="l", xlab="Time", ylim=c(-120,0), ylab="Root Tip Angle (degrees)", main="" )
for(i in c(19, 20, 132, 72))
  lines(1:241, pheno1[i,])
```
The estimated genetic
map and the trait values for five randomly selected RIL are displayed.


## LOD images

Now let's make an LOD image of all time points.
```{r scanone}
simspal <- calc.genoprob(simspal, step=1)
out <- scanone(simspal, pheno.col = 1:241, method="hk")
```
We first applied interval mapping by Haley-Knott regression (Haley
and Knott 1992), considering each time point individually.
The results are displayed below, with the y-axis representing genomic position
and the x-axis representing time, and so each vertical slice is a genome scan
for one time point.
```{r plot2lod, fig.height=10, fig.width = 12}

plot2lod(simspal, out, col=heat.colors(100)[100:1], main="The LOD
image of the simspal data set")
```
Figure shows LOD image of the data set. It looks like we have some QTL on chromosome 1 and 4. And maybe on 3.

We can also produce a signed lod image.
```{r signed lod}
simspal <- calc.genoprob(simspal, step=1)
out <- scanone(simspal, pheno.col=1:241 , method="hk")
eff <- geteffects(simspal, pheno.cols=1:241)
```

```{r plotload, fig.height=10, fig.width = 12}
plotlod(out, eff, gap=15, main="The LOD image of the simspal data set")
```
The results are displayed in the Figure above,
with the x-axis representing genomic position and the y-axis representing time, and so each horizontal slice is a genome scan for one time point. We plot a signed LOD score, with the sign representing the estimated direction of the QTL effect. The most prominant QTL are on chromosomes 1 and 4.

The chromosome 1 QTL affects later times, and the chromosome 4
allele affects earlier times. There is an additional QTL of interest on distal chromosome 3

## SLOD and MLOD scores

```{r slodmlod}
out1 <- scanoneF(simspal, pheno.cols = 1:241, method="hk")

```
The SLOD and MLOD statistics combine the results across time points, by taking the average
or the maximum LOD, respectively, at each genomic location.


```{r slodmlodcurve, fig.height=10, fig.width = 12}
par(mfrow=c(2,1))
plot(out1, ylim=c(0,3.5), main="The SLOD curve for IBMset1", bandcol="gray90")
abline(h=2.02, col="red", lty=3)
plot(out1[,c(1,2,4)], ylim=c(0,7), main="The MLOD curve for IBMset1", bandcol="gray90")
abline(h=3.46, col="red", lty=3)

# permutation threshold
# permout <- scanoneF(simspal, pheno.cols=1:241, method = "hk", n.perm=1000)
# summary(permout) # display 5, 10 % threshold of permutation result
```

The results are in Figure above. Horizontal lines indicate the 5% genome-wide significance thresholds, derived by a permutation test. We didn't run the code above since it takes a long time (about one hour maybe) .

## Getting multiple QTL

```{r stepwiseqtlscan}

#qtlslod <- stepwiseqtlF(simspal, pheno.cols = 1:241, max.qtl = 6,
#                       usec = "slod", method = "hk", penalties = c(2.02, 2.62, 1.74) )
simspal <- calc.genoprob(simspal, step=0)
qtlslod <- makeqtl(simspal, chr = c(1, 4),
               pos = c(36.6, 27.8), what = "prob")
```

`stepwiseqtlF` function returns qtl object by using stepwise QTL selection. It takes long time to run this function, so I run `makeqtl` function to get the result of `stepwiseqtlF` function. You will get the same result if you run `stepwiseqtlF` function. 


``` {r profilelodimage, fig.height = 10, fig.width = 14}
lodmat1.c <- getprofile(simspal, qtl =  qtlslod, pheno.cols =1:241,
                             formula = y~Q1 + Q2 , method = "hk",
							  verbose = F, tpy="comb")
plotprofile(lodmat1.c, mval = 8, col=heat.colors(100)[100:1], main="The Profile LOD image of data")
```
You can get profilelod image by using `getprofile` function and `plotprofile` function. You need to have an option `step = 0` in `calc.genoprob` function.

## Effect plot

```{r effectplot}

slodeff <- vector("list", nphe(simspal))

for(i in 1:nphe(simspal)) {
    slodeff[[i]] <- summary(fitqtl(simspal, phe=i, qtl=qtlslod,
                            method="hk", get.ests=TRUE, dropone=FALSE))$ests[,1]*c(1,2,2)
}

nam <- names(slodeff[[1]])
slodeff <- matrix(unlist(slodeff), byrow=TRUE, ncol=length(nam))
colnames(slodeff) <- nam

time <- (0:240)/30
```
To further characterize the effects of the QTL in the context of the inferred multiple-QTL
models, we fit the selected multiple-QTL models at each time point, individually. the estimated baseline function and the estimated QTL effects, as a function of time.
The estimated QTL effects in panels are for the difference between two alleles. 


```{r effplot, fig.height=5, fig.width = 14}
par(mfrow=c(1,3))
plot(time, slodeff[,1], lwd=2, type="l",
     xlab="Time (hours)",
	      ylab="Tip angle (degrees)", col="red", ylim=c(-110,0))
		  mtext("baseline curve", side=3, line=0.5)

plot(time, slodeff[,2], lwd=2, ylim = c(-5,9), type="l",
     xlab="Time (hours)",
	      ylab="QTL effect", col="red")
		  abline(h=0)
		  mtext("chr 1, 37 cM", side=3, line=0.5)

plot(time, slodeff[,3], lwd=2, ylim = c(-5,9), type="l",
     xlab="Time (hours)",
	      ylab="QTL effect", col="red")
		  mtext("chr 4, 28 cM", side=3, line=0.5)
		  abline(h=0)
```


